### DISCUSSION TITLE
fastapi for prod to support 100 requests

### DISCUSSION THREAD

**Discussion by PriyatamNayak** [fastapi/fastapi#9092] [Questions] [answered]
How can I set up this fastapi which can sustain 100 requests per second

Here the task is cpu intensive which took 10 seconds for every request

As we are running lp optimisation using ortools

Please help me

Infra

8 core cpu
Have used gunicorn with uvicornworker11  -w 8       
but it is not able to sustain more than 10 requests

**Comment by jorgerpo** [comment]
did you remove  the --reload flag?

**Comment by PriyatamNayak** [comment]
have not using it
am using the below
gunicorn example:app -w 8 -k uvicorn.workers.UvicornWorker

**Comment by GabrielCarpr** [comment]
If it's CPU intensive, and the task takes 10 seconds, does it really need to be run during the request? Can you not run it as a background task, or put it in a message queue and use Celery, for example?

**Comment by PriyatamNayak** [comment]
@GabrielCarpr 
Because other api (engine) which is built on java needs this result do the calculation and give results to the user
flow is like this:   user triggers some calculation by giving some data,then it call another engine to do the calc after it gets results from another engine , it will call this python api to do futher calculation...then main engine collects data and do some calc and give results to users

so total is taking 30 seconds...for small data is taking  8 seconds 

this api for samll data is less than 1second

**Comment by sm-Fifteen** [comment, marked as answer]
The most common cause of performance drops for CPU-heavy requests tends to be stalled event loops and GIL contention. You haven't shown any code yet, so I'll guess it's the former. Can you try setting `PYTHONASYNCIODEBUG=1` in your environment variables? Something like this, maybe (all in one line, that's how you set [temporary environment variables in bash](https://www.gnu.org/software/bash/manual/html_node/Environment.html)):

`PYTHONASYNCIODEBUG=1 gunicorn example:app -w 8 -k uvicorn.workers.UvicornWorker`

If you're doing heavy processing, depending on how everything is configured on your end, it's possible that the processing is preventing the event loop from getting back to listening for new connections. If the above command starts giving you warnings about blocking calls taking too long, it means your event loop is getting stalled like this. It's a common mistake that has to do with [how FastAPI handles sync and async routes](https://fastapi.tiangolo.com/async/#very-technical-details). You usually want to avoid doing much processing on your API server directly like that and stick to I/O, which is why most people don't really mention that.

**Comment by Miguelme** [comment]
Did it work ? I am also stuck at 10-12 request per second with a similar case.

**Comment by PriyatamNayak** [comment]
NO

**Comment by Mause** [comment]
Can you provide a minimal code reproduction so that we can help you?

**Comment by vcidst** [comment]
Will the answer here be different if instead of a CPU heavy task it was one(or few) network calls?

```
from fastapi import FastAPI
import requests

app = FastAPI()

# takes at least 10 seconds to respond
# example of a long polling endpoint
@app.post("/webhook")
def webhook():
  requests.get("http://slowwly.robertomurray.co.uk/delay/10000/url/http://www.google.com")
  return {"Status": "OK"}
```

I deployed this with Uvicorn and Gunicorn,
```
gunicorn main:app -w 4 -k uvicorn.workers.UvicornWorker
```

However the load tests from `hey` show that this server is unable to serve 10+ requests per second

**Comment by PriyatamNayak** [comment]
Haha ..same in my case..but you can try with async def instead of def

**Comment by vgibilmanno** [comment]
I'm hitting the same problem @vcidst were you able to find a solution for this?

**Comment by PriyatamNayak** [comment]
Any one got work around?

**Comment by sm-Fifteen** [comment]
> Haha ..same in my case..but you can try with async def instead of def

`async def` tells FastAPI the function is  not expected to block, but `requests` does block when waiting for a response. That means your entire worker thread will be waiting for your function to finish or `await` something before it moves on to handling any other request.

If that route then blocks for 10 seconds and you have 4 worker threads, it would make sense that the most you can handle under that setup is 4 requests per 10 seconds. You should either change that route to be `def` so FastAPI knows to spin it in its own thread and not block the worker, or use an async http library like `httpx` so you can `await` and free up your worker while waiting for the response to come in.

> Can you try setting `PYTHONASYNCIODEBUG=1` in your environment variables? Something like this, maybe (all in one line, that's how you set [temporary environment variables in bash](https://www.gnu.org/software/bash/manual/html_node/Environment.html)):
> 
> `PYTHONASYNCIODEBUG=1 gunicorn example:app -w 8 -k uvicorn.workers.UvicornWorker`

@PriyatamNayak: Did you try this? It's usually very helpful in figuring out why an async python application is blocking.

**Comment by ycd** [comment]
There is no way to reproduce your issue @PriyatamNayak.  We can't help with just "assuming" what your code looks like.

**Comment by vgibilmanno** [comment]
@PriyatamNayak in my case the performance issues came from the application being too CPU heavy causing the GIL to run the application in a single threaded manner than multi threaded. @sm-Fifteen mentioned that in a previous post already but at that time I didn't know about the GIL and what it did.

**Comment by gaurav-95** [comment]
I'm running a spacy library, continuously receiving text as payload from node.js server. for each text im processing and returning using fastapi. The api works well for first few minutes then slows down as starts rejecting requests. Any way threading can help me?

**Comment by matanmarciano** [comment]
Same here, any news?
